src/main/java/picard/analysis/CompareMetrics.java

```diff
@@ -50,35 +70,244 @@
+                    errs.add("Invalid value for Argument 'METRIC_ALLOWABLE_RELATIVE_CHANGE'");
+                }
+            }
+        }
+
+        if (errs.isEmpty()) {
+            return null;
+        } else {
+            return errs.toArray(new String[0]);
+        }
+    }
+
+    private int compareMetricsFiles(final File metricFile1, final File metricFile2) throws IOException, IllegalAccessException {
+        final MetricsFile<?, ?> mf1 = new MetricsFile<>();
+        final MetricsFile<?, ?> mf2 = new MetricsFile<>();
+        mf1.read(new FileReader(metricFile1));
+        mf2.read(new FileReader(metricFile2));
+        if (mf1.areMetricsEqual(mf2)) {
+            return 0;
+        } else if (!mf1.getMetrics().get(0).getClass().equals(mf2.getMetrics().get(0).getClass())) {
+            throw new PicardException("Metrics are of differing class between " + metricFile1.getAbsolutePath() + " and " + metricFile2.getAbsolutePath());
+        } else if (!mf1.getMetricsColumnLabels().equals(mf2.getMetricsColumnLabels())) {
+            log.error("Metric columns differ between " + metricFile1.getAbsolutePath() + " and " + metricFile2.getAbsolutePath());
+            return 1;
+        } else if (mf1.getMetrics().size() != mf2.getMetrics().size()) {
+            log.error("Number of metric rows differ between " + metricFile1.getAbsolutePath() + " and " + metricFile2.getAbsolutePath());
+            return 1;
+        }
+
+        validateMetricNames(mf1, metricFile1, METRICS_TO_IGNORE);
+        validateMetricNames(mf1, metricFile1, MetricToAllowableRelativeChange.keySet());
+
+        Set<String> metricsToIgnore = new HashSet<>(METRICS_TO_IGNORE);
+
+        final Class<? extends MetricBase> metricClass = mf1.getMetrics().get(0).getClass();
+
+        int retVal = 0;
+        int rowNumber = -1;
+        final Field[] fields = metricClass.getFields();
+        Iterator<?> mf1Iterator = mf1.getMetrics().iterator();
+        Iterator<?> mf2Iterator = mf2.getMetrics().iterator();
+        while (mf1Iterator.hasNext() && mf2Iterator.hasNext()) {
+            rowNumber++;
+            MetricBase metric1 = (MetricBase) mf1Iterator.next();
+            MetricBase metric2 = (MetricBase) mf2Iterator.next();
+            for (Field field : fields) {
+                if (!metricsToIgnore.contains(field.getName())) {
+                    final Object value1 = field.get(metric1);
+                    final Object value2 = field.get(metric2);
+                    SimpleResult result = compareMetricValues(value1, value2, field.getName());
+                    if (!result.equal) {
+                        retVal = 1;
+                        final String diffString = "Row: " + rowNumber + " Metric: " + field.getName() +
+                                " values differ. Value1: " + value1 + " Value2: " + value2 + " " + result.description;
+                        differences.add(diffString);
+                        log.error(diffString);
```

Comment:
Is it important to log when adding, vs logging at the end when all differences have been generated? If you log at the end, then you don't need to log both here and below with the histogram check.

Reply:
that's true.